adr/ADR-0001-trustyai-external-library-integration.md

```diff
@@ -0,0 +1,114 @@
+* The Java service will execute the Python algorithms by invoking the API wrapper and passing all the necessary data, read from storage.
+* Python results will be converted into Java data structures
+* Results will be returned by the service (either as responses to ad-hoc requests, or as a Prometheus data). This step is handled exclusively by the Java part. As such, the external communication is delegated to the service.
+
+The API wrapper would provide a general enough abstraction for explainers and metrics so that they could be treated by the TrustyAI service as a regular Java implementation. Any Python implementation-specific initialization or configuration would be implemented on the wrapper side.
+jep provides specific optimizations for the serialization/deserialization of numpy data (in the form of Java types `NDFloat`, etc) which could be leveraged in this scenario.
+
+![](assets/ADR-0001-a.png)
+
+**Pros:**
+
+* External algorithms co-located with TrustyAI service
+* Fewer external calls / network communication
+
+**Cons:**
+
+* Increase complexity of TrustyAI service (dependency tracking and updating)
+* Adding new algorithms requires hardcoding
+* Reduced flexibility (third party libraries need to be added directly)
+
+### Proposal 2 - Container/Microservice composition: Sidecar container
+
+This proposal is based on [KServe’s approach to model explainability](https://kserve.github.io/website/0.8/modelserving/explainer/explainer/#inferenceservice-explainer). We assume the case of Python external libraries which need to be used as the algorithmic backend for the TrustyAI service.
+
+* A Python base class is provided, which implements the intra-service communication transparently (e.g. an “explain” method, a “register” method, etc).
+  * Communication between TrustyAI and external service implemented in gRPC
+* External Python algorithms are implemented as a subclass and packaged as a container.
+* When the external service is deployed, it automatically registers its capabilities with a TrustyAI service, providing
+  * The external service’s location
+  * The external service’s capabilities (algorithm names, TrustyAI endpoint claim)
+* A request to TrustyAI for an explainer/metric matching the externally registered ones will be invoked by
+  * Passing data to the external service
+  * External service executes computations
+  * Final result is returned to TrustyAI
+* TrustyAI published result to Prometheus or return response if it’s an ad-hoc request
+
+![](assets/ADR-0001-b.png)
+![](assets/ADR-0001-c.png)
+![](assets/ADR-0001-d.png)
+
+**Pros:**
+
+* Extensible for any external service that implements the API
+
+**Cons:**
+
+* Limitations of passing large amount of data through the gRPC channel
+* Communication API should accommodate different explainer/metric APIs
+* Write gRPC client/service protobuf
```

Comment:
I would make more explicit that this proposal requires some more work on third-party side to implement this gRPC API. Wdyt?

```suggestion
* Each library should implement the API writing a gRPC client/service protobuf to communicate
```

Reply:
Fine for me 👍